### PR TITLE
Make bevy_math optional in bevy_core

### PR DESCRIPTION
**Pull request by CAD97**
Notably, this makes bevy_core, bevy_diagnostic, and bevy_asset no longer require pulling in bevy_math. None of these crates directly depend on the math primitives, so this is a net positive for use outside of bevy, which can now use these crates without pulling in bevy_math and glam.

# Objective

I'm building my own micro-engine off of the bevy primitives for learning purposes, and as part of that, I'm using my own simplistic math library rather than using bevy_math/glam. (Basically, I would like to have everything up to but not including rendering and math.) I was looking at pulling in `bevy_asset` to handle asset loading (an ugly task I have no interest in re-solving), but noticed it was pulling in `bevy_math` indirectly.

Cue surprise that `bevy_math` isn't actually _used_ in any fashion, just to eagerly `.register_type` in `CorePlugin`. Putting that functionality behind `features = ["bevy"]` (as is done in `bevy_reflect`) is a very minor change that enables me to use `bevy_asset` without needlessly pulling in glam.

## Solution

- Gate `CorePlugin`'s registering of `bevy_math` types for reflection behind an optional `bevy_math` dependency
- Add a `bevy` feature to `bevy_core` which enables `bevy_math` and `bevy_reflect`'s `bevy` feature
- Enable `bevy_core`'s `bevy` feature via `bevy_internal`
  - This is the decision which is most arbitrary, but doing so felt most natural; the feature is necessarily only used by the consumer of `CorePlugin`, which is the `bevy` end-user.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
Looks good! Any detangling of our dependency bush is a win IMO.

The code changes here are very simple.

**Review by inodentry — APPROVED** [review]
Yes, this seems like an easy win for bevy's modularity and usefulness in other projects, without significant drawback for use inside bevy. I like this. :)

**Review by mockersf — APPROVED** [review]

**Review by IceSentry — APPROVED** [review]

**Comment by CAD97** [comment]
I think the core-ish conflict here is that `bevy_core` is sort of doing two, maybe three distinct jobs. It's providing the `name` and `time` modules' types for use in an ECS, `label` module for use with bevy_ecs, the `Bytes` interface for GPU buffer serialization, and the `CorePlugin` for bevy_app for initializing the above resources, initializing bevy_task, and registering primitive and bevy_math types for bevy_serialization.

The "problem" then comes from pulling one necessarily pulling the others. bevy_diagnostic only uses the time types, and bevy_asset even only optionally reports any diagnostics.

My "issue" isn't that bevy_core assumes bevy_math is desirable, it's more that bevy_asset, which _is_ quite individually useful, has a dep noticeably fatter tree than required for standalone use with bevy_app/bevy_ecs but not the whole of bevy.

It's fine if this is a modularity border that the bevy project doesn't see value in setting -- after all, the difference is just cold compile times for downstream using bevy piecewise, since the unused code/crates will trivially optimize out -- but I personally think that this is a border worth putting at least a little bit of "as-is" support into, thus filing this PR.

But, again, it's perfectly understandable if the bevy project doesn't want to carve this out this way. And if there _is_ a better way that you'd like to provide bevy_asset for lighterweight use, I'd be happy to implement that approach as well.

<details><summary>Trimmed, annotated dep tree of bevy_asset</summary>

```text
bevy_asset v0.5.0 (a60fe30)
├── bevy_app v0.5.0 (a60fe30)
│   ├── bevy_derive v0.5.0 (proc-macro) (a60fe30) (...)
│   ├── bevy_ecs v0.5.0 (a60fe30) (...)
│   ├── bevy_reflect v0.5.0 (a60fe30) (*)
│   └── bevy_utils v0.5.0 (a60fe30) (*)
├── bevy_diagnostic v0.5.0 (a60fe30) (?)
│   ├── bevy_app v0.5.0 (a60fe30) (*)
│   ├── bevy_core v0.5.0 (a60fe30)
│   │   ├── bevy_app v0.5.0 (a60fe30) (*)
│   │   ├── bevy_derive v0.5.0 (proc-macro) (a60fe30) (*)
│   │   ├── bevy_ecs v0.5.0 (a60fe30) (*)
│   │   ├── bevy_math v0.5.0 (a60fe30) (..!)
│   │   ├── bevy_reflect v0.5.0 (a60fe30) (*)
│   │   ├── bevy_tasks v0.5.0 (a60fe30) (*)
│   │   ├── bevy_utils v0.5.0 (a60fe30) (*)
│   │   └── ... (*)
│   ├── bevy_ecs v0.5.0 (a60fe30) (*)
│   ├── bevy_log v0.5.0 (a60fe30)
│   └── bevy_utils v0.5.0 (a60fe30) (*)
├── bevy_ecs v0.5.0 (a60fe30) (*)
├── bevy_log v0.5.0 (a60fe30) (*)
├── bevy_reflect v0.5.0 (a60fe30) (*)
├── bevy_tasks v0.5.0 (a60fe30) (*)
├── bevy_utils v0.5.0 (a60fe30) (*)
└── ...
```

Symbol key: `*`: included previously; `...`: omitted (transitive) deps; `?`: questionable/optional/minimal/noncore use (to root); `!`: basically unused (to root)
</details>

**Comment by cart** [comment]
I do agree that this is worth discussing and I think you've zero-ed in on the core issue: `bevy_core` currently serves a number of different roles:

* type registrations: math / std / ecs / time / misc
* time
* names / labels
* FloatOrd
* Byte conversion
* task pool setup

It is the "glue" that ties miscellaneous things without their own crates into bevy apps, often when the actual crates cannot do the registering themselves due to circular dependencies (ex: bevy_tasks is used in bevy_ecs, so we can't handle plugin stuff in bevy_tasks). I do think that it is worth revisiting bevy_core. I object to the current impl in this pr for the reasons I called out above, but I think we can probably find a solution that satisfies everyone:

* Byte conversion will be deprecated in favor of bytemuck and crevice once the new renderer gets merged
* bevy_math might actually be able to define its own plugin / take a dependency on bevy_app. this forces bevy_math consumers to also consume bevy_app / bevy_ecs / etc, but maybe thats ok.
* time stuff could be broken out into bevy_time
* FloatOrd could be moved to bevy_utils (although its type registration cannot)
* Type registrations could be moved somewhere else (maybe even the top level `bevy_internal` crate)
* We could consider doing a task pool setup plugin in bevy_ecs or bevy_internal

In short, we might be able to do away with bevy_core entirely. I'd prefer to start doing this stuff after the new renderer is merged to avoid conflicts / allow us to remove the Bytes stuff entirely. These are all straightforward changes that won't take any real time investment.
